admin-api-server/src/nft/service/file/file.service.ts

```diff
@@ -0,0 +1,460 @@
+import { Injectable, Logger } from '@nestjs/common';
+import fs from 'fs';
+import { Buffer } from 'node:buffer';
+import ffmpeg from 'fluent-ffmpeg';
+import { fileURLToPath } from 'url';
+import * as path from 'path';
+import sharp from 'sharp';
+import sizeOf from 'image-size';
+import { v4 as uuidv4 } from 'uuid';
+import mime from 'mime';
+
+mime.define(
+  {
+    'video/mov': ['mov'],
+    'video/quicktime': ['qt'],
+  },
+  true,
+);
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+
+export interface File {
+  originalname: string;
+  fieldname: string;
+  mimetype: string;
+  encoding: string;
+  buffer: Buffer;
+  size: number;
+}
+
+type FileType = 'thumbnail' | 'display';
+type Path = string;
+type Dimension = {
+  width: number;
+  height: number;
+};
+
+interface ResizedImgParams {
+  buffer: Buffer;
+  imageSize: { width: number; height: number };
+  type: FileType;
+  extension: string;
+}
+
+interface ImageFromVideoParams {
+  videoFile: File;
+  typeOfImage: FileType;
+  imageSize: number;
+}
+
+interface GenerateImageFileParams {
+  source: Path | Buffer;
+  type: FileType;
+  extension: string;
+}
+
+interface ScreenshotFromVideoParams {
+  imageSize: number;
+  videoPath: string;
+  filename: string;
+}
+
+interface HandleVideoParams {
+  thumbnailMissing: boolean;
+  displayMissing: boolean;
+}
+
+interface HandleImageParams {
+  thumbnailMissing: boolean;
+  displayMissing: boolean;
+}
+
+@Injectable()
+export class FileService {
+  private MEDIA_TEMP_DIR: string;
+  private allFiles: File[] = [];
+  private artifact: File;
+
+  async addMissingFiles(filesArray: File[]): Promise<File[]> {
+    this.MEDIA_TEMP_DIR = __dirname + `/${uuidv4()}/`;
+    this.allFiles = filesArray;
+    this.artifact = filesArray.find(
+      (file) => file.originalname === 'artifact',
+    ) as File; // we always assume that we get an artifact from the client
+
+    const thumbnailMissing = !filesArray.some(
+      (file) => file.originalname === 'thumbnail',
+    );
+    const displayMissing = !filesArray.some(
+      (file) => file.originalname === 'display',
+    );
+    const isVideo = (this.artifact.mimetype.match(/video/g) || []).length > 0;
+    const isImage = (this.artifact.mimetype.match(/image/g) || []).length > 0;
+
+    try {
+      if (isVideo) {
+        this.#createTempDir();
+        const addedFiles = await this.#handleVideo({
+          thumbnailMissing,
+          displayMissing,
+        });
+
+        if (addedFiles.length) {
+          addedFiles.forEach((file) => {
+            this.allFiles.push(file);
```

Comment:
The implementation here shouldn't use a global class instance variable for accumulating state.

1. this will not work concurrently where multiple users are uploading an Nft at the same time, because both will hit this same global state
2. even if I am wrong about point 1, it is much more difficult to reason about global state than it is about local state. And we're really not gaining anything here by using global class state. Much better to accumulate the result of this function locally with a `res: File[]` variable.